xapian-core/tests/api_weight.cc

```diff
@@ -882,6 +882,39 @@ DEFINE_TESTCASE(tfidfweight3, backend) {
     mset_expect_order(mset, 2, 4);
     TEST_EQUAL_DOUBLE(mset[0].get_weight(), 8 * log((6.0 - 2) / 2));
     TEST_EQUAL_DOUBLE(mset[1].get_weight(), 1 * log((6.0 - 2) / 2));
+
+    // Check for "ann".
+    enquire.set_weighting_scheme(Xapian::TfIdfWeight("ann"));
+    mset = enquire.get_mset(0, 10);
+    TEST_EQUAL(mset.size(), 2);
+    mset_expect_order(mset, 2, 4);
+    TEST_EQUAL_DOUBLE(mset[0].get_weight(), 0.2 + 0.8 * log(1.0 + 8));
+    TEST_EQUAL_DOUBLE(mset[1].get_weight(), 0.2 + 0.8 * log(1.0 + 1));
+
+    // Check for "nGn" .
+    enquire.set_weighting_scheme(Xapian::TfIdfWeight("nGn"));
+    mset = enquire.get_mset(0, 10);
+    TEST_EQUAL(mset.size(), 2);
+    mset_expect_order(mset, 2, 4);
+    TEST_EQUAL_DOUBLE(mset[0].get_weight(), 8 * (9.0 / 2));
+    TEST_EQUAL_DOUBLE(mset[1].get_weight(), 1 * (9.0 / 2));
+
+    // Check for "Snn" .
+    enquire.set_weighting_scheme(Xapian::TfIdfWeight("Snn"));
+    mset = enquire.get_mset(0, 10);
+    TEST_EQUAL(mset.size(), 2);
+    mset_expect_order(mset, 2, 4);
+    TEST_EQUAL_DOUBLE(mset[0].get_weight(), sqrt(8 - 0.5) + 1);
+    TEST_EQUAL_DOUBLE(mset[1].get_weight(), sqrt(1 - 0.5) + 1);
+
+    // Check for "nln" .
+    enquire.set_weighting_scheme(Xapian::TfIdfWeight("nln"));
+    mset = enquire.get_mset(0, 10);
+    TEST_EQUAL(mset.size(), 2);
+    mset_expect_order(mset, 2, 4);
+    TEST_EQUAL_DOUBLE(mset[0].get_weight(), 8 * log(9.0 / 2 + 1));
+    TEST_EQUAL_DOUBLE(mset[1].get_weight(), 1 * log(9.0 / 2 + 1));
```

Comment:
There's no test of 'E' for the second character here (which would probably have revealed that you didn't actually provide an implementation for it).

Reply:
'E' is for Entropy normalization . I haven't implemented it yet .
It requires visibility of wdf of the term in every document in the collection . The idea was to use the postlist to get that.
But the concern is that won't be very efficient.
So it is still to be implemented

Reply:
I will make a separate PR for Entropy normalization.

Reply:
Good plan.